src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java

```diff
@@ -267,6 +289,17 @@ public MqttClient publish(String topic, Buffer payload, MqttQoS qosLevel, boolea
 
     io.netty.handler.codec.mqtt.MqttMessage publish = MqttMessageFactory.newMessage(fixedHeader, variableHeader, buf);
 
+    switch (qosLevel) {
+      case AT_LEAST_ONCE:
+        qos1outbound.put(variableHeader.messageId(), publish);
+        break;
+      case EXACTLY_ONCE:
+        qos2outbound.put(variableHeader.messageId(), publish);
+        break;
+    }
+
+    countInflightQueue++;
```

Comment:
incrementing here, we have that it counts the QoS = 0 message as well which won't go into any queue. It's not right to increment this counter here.